### PR TITLE
fix max backward underflow for float16

### LINKED ISSUE #12296: max backward can underflow
```
from tinygrad import Tensor
t = Tensor.ones(70000, dtype="half", requires_grad=True).contiguous()
t.max().backward()
print(t.grad.tolist()[:10])
```
gives `[0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0]`

https://github.com/tinygrad/tinygrad/blob/476a2a0a962fa8a145c612f63b72eeb4dc3a84f3/tinygrad/gradient.py#L10 uses the gradient dtype for counting

### PR DESCRIPTION
**Pull request by mango766**
Fixes #12296

`Tensor.ones(70000, dtype="half").max().backward()` returns all-zero gradients because the mask counting in `reduce_gradient` uses the gradient dtype (float16) for accumulation. Float16 can only represent integers up to 2048 exactly, so the count of 70000 matching elements overflows and the subsequent division produces zeros.

**Fix:** Use `sum_acc_dtype` (already used throughout the codebase for safe accumulation, e.g. in `Tensor.sum` and `Tensor.mean`) to upcast the mask before counting, then cast back to the gradient dtype afterward.

Before:
```
>>> Tensor.ones(70000, dtype="half", requires_grad=True).contiguous().max().backward()
>>> t.grad.tolist()[:5]
[0.0, 0.0, 0.0, 0.0, 0.0]
```

After:
```
>>> t.grad.tolist()[:5]
[1.430511474609375e-05, 1.430511474609375e-05, 1.430511474609375e-05, 1.430511474609375e-05, 1.430511474609375e-05]
```

### CODE REVIEW

**Comment by geohot** [comment]
Hmm, this is a pretty big change. What is it currently doing? What does torch do here?
